Upsetplot

wdecoster opened the pull request:
This is my first PR to add a recipe to conda forge. The package I am adding a recipe for is in Python.

Checklist

- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x] Source is from official source
- [x] Package does not vend other packages
- [x] Build number is 0
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/upsetplot```) and found some lint.

Here's what I've got...


For **recipes/upsetplot**:

 * There are 1 too many lines.  There should be one empty line at the end of the file.

scopatz commented on recipes/upsetplot/meta.yaml:
```diff
@@ -0,0 +1,43 @@
+extra:
+  recipe-maintainers:
+    - wdecoster
+
```
You need to remove this line :wink:

wdecoster replied:
Ah, will try that! The linter seemed to suggest I needed to add one, seems I misunderstood.

scopatz commented on recipes/upsetplot/meta.yaml:
```diff
@@ -0,0 +1,43 @@
+
+about:
+  home: https://github.com/jnothman/UpSetPlot
+  license: BSD
```
Which version of the BDS license?

wdecoster replied:
I've specified now "New BSD License". Is that okay as such?

wdecoster replied:
Seems it cannot include the word "License" according to the linter. Let's try just "New BSD" then.

scopatz replied:
Yeah, "BSD-3-Clause" is usually used.

wdecoster replied:
Alright, will change.

scopatz commented on recipes/upsetplot/meta.yaml:
```diff
@@ -0,0 +1,43 @@
+about:
+  home: https://github.com/jnothman/UpSetPlot
+  license: BSD
+  license_family: BSD
```
This recipe needs a `license_file` entry

wdecoster replied:
`license_file: LICENSE` will do, for a file called `LICENSE`? Or is a .txt extension prefered?

scopatz replied:
no extension needed.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/upsetplot```) and found some lint.

Here's what I've got...


For **recipes/upsetplot**:

 * The recipe `license` should not include the word "License".

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/upsetplot```) and found it was in an excellent condition.

wdecoster commented:
Thanks for the feedback @scopatz!

wdecoster commented:
I must be missing a pip install somewhere, as my recipe does not specify to install upsetplot itself?

scopatz commented on recipes/upsetplot/meta.yaml:
```diff
@@ -0,0 +1,40 @@
+  license: New BSD
+  license_family: BSD
+  summary: 'Creating upsetplots in Python'
+  license_file: LICENSE
```
this is out of order, please move below license_family and above summary

wdecoster replied:
Fixed!

scopatz commented on recipes/upsetplot/meta.yaml:
```diff
@@ -0,0 +1,40 @@
+
+build:
+  noarch: python
+  number: 0
```
you need to have the script line from the example/meta.yaml here

wdecoster replied:
Of course :grin: :man_facepalming:  Thanks!

wdecoster commented:
I see `DEPRECATION: Python 2.7 will reach the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 won't be maintained after that date. A future version of pip will drop support for Python 2.7.`? 

And also:

```
Running setup.py (path:/tmp/pip-req-build-W37urO/setup.py) egg_info for package from file://$SRC_DIR
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-req-build-W37urO/setup.py", line 7, in <module>
        from upsetplot import __version__ as version
      File "upsetplot/__init__.py", line 1, in <module>
        from .plotting import UpSet, plot
      File "upsetplot/plotting.py", line 9, in <module>
        from matplotlib import pyplot as plt
      File "/home/conda/feedstock_root/build_artifacts/upsetplot_1562072347034/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib/python2.7/site-packages/matplotlib/pyplot.py", line 115, in <module>
        _backend_mod, new_figure_manager, draw_if_interactive, _show = pylab_setup()
      File "/home/conda/feedstock_root/build_artifacts/upsetplot_1562072347034/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib/python2.7/site-packages/matplotlib/backends/__init__.py", line 63, in pylab_setup
        [backend_name], 0)
      File "/home/conda/feedstock_root/build_artifacts/upsetplot_1562072347034/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib/python2.7/site-packages/matplotlib/backends/backend_tkagg.py", line 4, in <module>
        from . import tkagg  # Paint image to Tk photo blitter extension.
      File "/home/conda/feedstock_root/build_artifacts/upsetplot_1562072347034/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib/python2.7/site-packages/matplotlib/backends/tkagg.py", line 5, in <module>
        from six.moves import tkinter as Tk
      File "/home/conda/feedstock_root/build_artifacts/upsetplot_1562072347034/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib/python2.7/site-packages/six.py", line 203, in load_module
        mod = mod._resolve()
      File "/home/conda/feedstock_root/build_artifacts/upsetplot_1562072347034/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib/python2.7/site-packages/six.py", line 115, in _resolve
        return _import_module(self.mod)
      File "/home/conda/feedstock_root/build_artifacts/upsetplot_1562072347034/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib/python2.7/site-packages/six.py", line 82, in _import_module
        __import__(name)
      File "/home/conda/feedstock_root/build_artifacts/upsetplot_1562072347034/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib/python2.7/lib-tk/Tkinter.py", line 39, in <module>
        import _tkinter # If this fails your Python may not be configured for Tk
    ImportError: libX11.so.6: cannot open shared object file: No such file or directory
```

scopatz commented on recipes/upsetplot/meta.yaml:
```diff
@@ -0,0 +1,41 @@
+    - python
+    - pip
+    - pandas
+    - matplotlib >=2.0
```
Do you really need matplotlib and pandas to be able to build?

wdecoster replied:
Probably not!

wdecoster commented:
Puzzling build error:

```
Solving environment: ...working... done
Traceback (most recent call last):
  File "/home/conda/.ci_support/build_all.py", line 131, in <module>
    build_all(args.recipes_dir, args.arch)
  File "/home/conda/.ci_support/build_all.py", line 68, in build_all
    build_folders(recipes_dir, new_comp_folders, arch, channel_urls)
  File "/home/conda/.ci_support/build_all.py", line 121, in build_folders
    conda_build.api.build([recipe], config=get_config(arch, channel_urls))
  File "/opt/conda/lib/python3.7/site-packages/conda_build/api.py", line 209, in build
    notest=notest, need_source_download=need_source_download, variants=variants)
  File "/opt/conda/lib/python3.7/site-packages/conda_build/build.py", line 2313, in build_tree
    notest=notest,
  File "/opt/conda/lib/python3.7/site-packages/conda_build/build.py", line 1322, in build
    output_metas = expand_outputs([(m, need_source_download, need_reparse_in_env)])
  File "/opt/conda/lib/python3.7/site-packages/conda_build/render.py", line 738, in expand_outputs
    for (output_dict, m) in _m.copy().get_output_metadata_set(permit_unsatisfiable_variants=False):
  File "/opt/conda/lib/python3.7/site-packages/conda_build/metadata.py", line 2047, in get_output_metadata_set
    bypass_env_check=bypass_env_check)
  File "/opt/conda/lib/python3.7/site-packages/conda_build/metadata.py", line 721, in finalize_outputs_pass
    permit_unsatisfiable_variants=permit_unsatisfiable_variants)
  File "/opt/conda/lib/python3.7/site-packages/conda_build/render.py", line 519, in finalize_metadata
    exclude_pattern)
  File "/opt/conda/lib/python3.7/site-packages/conda_build/render.py", line 404, in add_upstream_pins
    permit_unsatisfiable_variants, exclude_pattern)
  File "/opt/conda/lib/python3.7/site-packages/conda_build/render.py", line 373, in _read_upstream_pin_files
    extra_run_specs = get_upstream_pins(m, actions, env)
  File "/opt/conda/lib/python3.7/site-packages/conda_build/render.py", line 361, in get_upstream_pins
    specs = _read_specs_from_package(loc, dist)
  File "/opt/conda/lib/python3.7/site-packages/conda/exports.py", line 198, in __call__
    value = self.func(*args, **kw)
  File "/opt/conda/lib/python3.7/site-packages/conda_build/render.py", line 256, in _read_specs_from_package
    if not specs and os.path.isfile(pkg_loc):
  File "/opt/conda/lib/python3.7/genericpath.py", line 30, in isfile
    st = os.stat(path)
TypeError: stat: path should be string, bytes, os.PathLike or integer, not NoneType
##[error]Bash exited with code '1'.
```

scopatz commented:
this is a conda-build issue

wdecoster commented:
I assume that means that's nothing I can fix?

scopatz commented:
Not per se.  I am rerunning the build now.  Let me know if it works

wdecoster commented:
It did change the error message:

```
Processing $SRC_DIR
  Created temporary directory: /tmp/pip-req-build-WyEPpM
  Added file://$SRC_DIR to build tracker '/tmp/pip-req-tracker-I06Yyn'
    Running command python setup.py egg_info
    Running setup.py (path:/tmp/pip-req-build-WyEPpM/setup.py) egg_info for package from file://$SRC_DIR
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-req-build-WyEPpM/setup.py", line 7, in <module>
        from upsetplot import __version__ as version
      File "upsetplot/__init__.py", line 1, in <module>
        from .plotting import UpSet, plot
      File "upsetplot/plotting.py", line 6, in <module>
        import numpy as np
    ImportError: No module named numpy
Cleaning up...
  Removing source in /tmp/pip-req-build-WyEPpM
Removed file://$SRC_DIR from build tracker '/tmp/pip-req-tracker-I06Yyn'
Removed build tracker '/tmp/pip-req-tracker-I06Yyn'
ERROR: Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-req-build-WyEPpM/
Exception information:
Traceback (most recent call last):
  File "$PREFIX/lib/python2.7/site-packages/pip/_internal/cli/base_command.py", line 178, in main
    status = self.run(options, args)
  File "$PREFIX/lib/python2.7/site-packages/pip/_internal/commands/install.py", line 352, in run
    resolver.resolve(requirement_set)
  File "$PREFIX/lib/python2.7/site-packages/pip/_internal/resolve.py", line 131, in resolve
    self._resolve_one(requirement_set, req)
  File "$PREFIX/lib/python2.7/site-packages/pip/_internal/resolve.py", line 294, in _resolve_one
    abstract_dist = self._get_abstract_dist_for(req_to_install)
  File "$PREFIX/lib/python2.7/site-packages/pip/_internal/resolve.py", line 242, in _get_abstract_dist_for
    self.require_hashes
  File "$PREFIX/lib/python2.7/site-packages/pip/_internal/operations/prepare.py", line 362, in prepare_linked_requirement
    abstract_dist.prep_for_dist(finder, self.build_isolation)
  File "$PREFIX/lib/python2.7/site-packages/pip/_internal/operations/prepare.py", line 171, in prep_for_dist
    self.req.prepare_metadata()
  File "$PREFIX/lib/python2.7/site-packages/pip/_internal/req/req_install.py", line 537, in prepare_metadata
    self.run_egg_info()
  File "$PREFIX/lib/python2.7/site-packages/pip/_internal/req/req_install.py", line 615, in run_egg_info
    command_desc='python setup.py egg_info')
  File "$PREFIX/lib/python2.7/site-packages/pip/_internal/utils/misc.py", line 776, in call_subprocess
    % (command_desc, proc.returncode, cwd))
InstallationError: Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-req-build-WyEPpM/
Traceback (most recent call last):
  File "/home/conda/.ci_support/build_all.py", line 131, in <module>
    build_all(args.recipes_dir, args.arch)
  File "/home/conda/.ci_support/build_all.py", line 68, in build_all
    build_folders(recipes_dir, new_comp_folders, arch, channel_urls)
  File "/home/conda/.ci_support/build_all.py", line 121, in build_folders
    conda_build.api.build([recipe], config=get_config(arch, channel_urls))
  File "/opt/conda/lib/python3.7/site-packages/conda_build/api.py", line 208, in build
    notest=notest, need_source_download=need_source_download, variants=variants)
  File "/opt/conda/lib/python3.7/site-packages/conda_build/build.py", line 2314, in build_tree
    notest=notest,
  File "/opt/conda/lib/python3.7/site-packages/conda_build/build.py", line 1477, in build
    cwd=src_dir, stats=build_stats)
  File "/opt/conda/lib/python3.7/site-packages/conda_build/utils.py", line 374, in check_call_env
    return _func_defaulting_env_to_os_environ('call', *popenargs, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/conda_build/utils.py", line 354, in _func_defaulting_env_to_os_environ
    raise subprocess.CalledProcessError(proc.returncode, _args)
subprocess.CalledProcessError: Command '['/bin/bash', '-e', '/home/conda/feedstock_root/build_artifacts/upsetplot_1562193209058/work/conda_build.sh']' returned non-zero exit status 1.
##[error]Bash exited with code '1'.
```

I believe this to be caused because the setup.py doesn't mention numpy? Although it does list pandas as a dependency, which I expect to be dependent on numpy. But maybe numpy has to be listed explicitly.

Pinging @jnothman

jnothman commented:
I can reproduce this! :|

wdecoster commented:
Do you think it would be fixed by having numpy as a dependency in your setup.py?

jnothman commented:
This was not a lack of dependency listing, but a problem with importing the version from `upsetplot`. I have fixed the import up, checked that `pipenv install .` works, tagged `0.3.0.post2` and twined it to PyPI.

jnothman commented:
Except that I made a silly mistake by rushing and will need to patch that with post3. Whoops.

wdecoster commented:
Gha, no worries. Let me know when it's up and I'll prepare the new recipe.

jnothman commented:
Travis says it's good now. 0.3.0.post3, thanks!!

wdecoster commented:
So it seemed also specifying `pytest-runner` as a dependency was necessary. Tests are successful now!
